core/amber/src/main/python/core/proxy/proxy_server.py

```diff
@@ -0,0 +1,237 @@
+            ```
+
+        :param original_func: decorated function, usually is a callable to be registered.
+        :param msg: the return message from the decorator, "ack" by default.
+        :return:
+        """
+
+        def ack_decorator(func: Callable):
+            @wraps(func)
+            def wrapper(*args, **kwargs):
+                func(*args, **kwargs)
+                return msg
+
+            return wrapper
+
+        if original_func:
+            return ack_decorator(original_func)
+        return ack_decorator
+
+    def __init__(self, scheme: str = "grpc+tcp", host: str = "localhost", port: int = 5005):
+        location = f"{scheme}://{host}:{port}"
+        super(ProxyServer, self).__init__(location)
+        logger.debug("Serving on " + location)
+        self.host = host
+
+        # actions for actions, will contain registered actions, identified by action name.
+        self._procedures: Dict[str, Tuple[Callable, str]] = dict()
+
+        # register heartbeat, this is the default action for the client to check the aliveness of the server
+        self.register("heartbeat", ProxyServer.ack()(lambda: None))
+
+        # register shutdown, this is the default action for the client to terminate the server.
+        self.register("shutdown",
+                      ProxyServer.ack(msg="Bye bye!")
+                      (lambda: threading.Thread(target=self._shutdown).start()),
+                      description="Shut down this server.")
+
+        self.register("control",
+                      ProxyServer.ack()(
+                          lambda control_message: self.process_control(control_message)),
+                      description="Process the control message"
+                      )
+
+        # the data message handler for each data message, needs to be implemented during runtime.
+        self.process_data = lambda *args, **kwargs: (_ for _ in ()).throw(NotImplementedError)
+
+        # the control message handler for each control message, needs to be implemented during runtime.
+        self.process_control = lambda *args, **kwargs: (_ for _ in ()).throw(NotImplementedError)
+
+    ###########################
+    # Flights related methods #
+    ###########################
+    @overrides(check_signature=False)
+    def do_put(self, context: ServerCallContext, descriptor: FlightDescriptor, reader: MetadataRecordBatchReader,
+               writer: RecordBatchStreamWriter):
+        """
+        put a data table into server, the data will be handled by the `self.process_data()` handler.
+        :param context: server context, containing information of middlewares.
+        :param descriptor: the descriptor of this batch of data.
+        :param reader: the input stream of batches of records.
+        :param writer: the output stream.
+        :return:
+        """
+
+        data: Table = reader.read_all()
+        logger.debug(f"getting a data batch {data}")
+        self.process_data(descriptor.command, data)
```

Comment:
Since `descriptor.command` could be arbitrary user-defined message, is there a reference of all possible types of `descriptor.command`?

Reply:
https://arrow.apache.org/docs/format/Flight.html according to Flight RPC proto, `descriptor.command` is bytes. In practice, it could be str or byte-like types.